### ISSUE TITLE
머신러닝 모델 클레스 정보

### ISSUE THREAD

**Issue by SeoYoungDeok** [AIPLE/AIPLE_WEB#1]
### CreateNN Class

|  No  |      Parameter      |              설명               |   Input_Type   |
| :--: | :-----------------: | :-----------------------------: | :------------: |
|  1   |     input_size      |         입력층 뉴런 수          |      int       |
|  2   |     hidden_size     |         은닉층 뉴런 수          |   int array    |
|  3   |     output_size     |         출력층 뉴런 수          |      int       |
|  4   |     activation      |           활성화 함수           |     string     |
|  5   |     w_init_std      | 가중치 초기화 [표준편차 / 방법] | float / string |
|  6   | weight_decay_lambda |        가중치 감소 세기         |     float      |
|  7   |       dropout       |       드롭아웃 사용 여부        |    boolean     |
|  8   |   dropout_ration    |          드롭아웃 비율          |     float      |
|  9   |      batchnorm      |      배치 정규화 사용 여부      |    boolean     |

##### w_init_std

* float : 가중치 초기화 시 표준편차를 입력 값으로 설정합니다.
* string : [he 초기값] 또는 [xavier 초기값]으로 설정합니다.
  * he 초기값 : relu 함수 사용 시 활성화 값을 고르게 분포시켜 줍니다.
  * xavier 초기값 : sigmoid 함수 사용 시 활성화 값을 고르게 분포시켜 줍니다.

------

##### weight_decay_lambda

* 큰 가중치에 대해 페널티를 부여함으로써 오버피팅을 억제합니다.

------

##### dropout

* 은닉층 뉴런을 무작위로 골라 신호 전달을 끊으며 학습하는 방법입니다. 오버피팅을 억제합니다.
* dropout_ration : dropout 비율을 설정 합니다.

------

##### batchnorm

* 배치 정규화 계층을 생성하여 각 층의 활성화값 분포가 고르게 퍼지도록 합니다.
* 학습 속도가 개선됩니다.
* 초기값의 의존도가 줄어듭니다.
* 오버피팅을 억제합니다.

### Trainer Class

|  No  |    Parameter    |        설명        |   Input_Type   |
| :--: | :-------------: | :----------------: | :------------: |
|  1   |     network     |   생성한 network   | CreateNN Class |
|  2   |     x_train     |    학습 데이터     |  numpy array   |
|  3   |     t_train     |  학습 정답 데이터  |  numpy array   |
|  4   |     x_test      |   테스트 데이터    |  numpy array   |
|  5   |     t_test      | 테스트 정답 데이터 |  numpy array   |
|  6   |     epochs      |   반복할 에폭 수   |      int       |
|  7   | mini_batch_size |  미니 배치 사이즈  |      int       |
|  8   |    optimizer    |    최적화 방법     |     string     |
|  9   | optimizer_param |    학습률 (lr)     |   dictionary   |

##### optimizer

* SGD : 기울기 방향으로 일정거리 만큼 이동하는 방법입니다.
* Momentum :  관성을 주어 일정 정도를 추가적으로 이동하는 방법입니다.
* Nesterov : Momentum을 기초로 한 방식으로 보다 효과적으로 이동하는 방법입니다.
* AdaGrad : 학습을진행하면서 학습률을 점차 줄여가는 방법입니다.
* RMSprop : AdaGrad가 학습을 오래 진행하면 갱신량이 0이 되는 문제를 개선한 방법입니다.
* Adam : Momentum과 RMSprop 방식을 합친 방법입니다.

### 예제 코드

```python
import sys, os
sys.path.append(os.pardir)  # 부모 디렉터리의 파일을 가져올 수 있도록 설정
import numpy as np
from CreateNN import *
from Trainer import *
import pickle
from dataset.mnist import load_mnist

(x_train, t_train), (x_test, t_test) = load_mnist(normalize=True, one_hot_label=True, flatten=True)

hidden_size = [256, 256, 256]

model = CreateNN(784, hidden_size, 10, activation='relu', w_init_std='relu', weight_decay_lambda=0, dropout=False, dropout_ration=0.1, use_batchnorm=True)

train = Trainer(model, x_train, t_train, x_test,  t_test, epochs=20, mini_batch_size=1000, optimizer='Adam', optimizer_param={'lr':0.01})

train.train()
```
